data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListener.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.expression;
+
+import com.amazon.dataprepper.model.event.Event;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ErrorNode;
+import org.antlr.v4.runtime.tree.TerminalNode;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionListener;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.AND;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.EQUAL;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.GT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.GTE;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.LT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.LTE;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT_EQUAL;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.OR;
+
+/**
+ * @since 1.3
+ * This listener implements {@link DataPrepperExpressionListener} to provide callbacks to handle evaluation of
+ * {@link org.antlr.v4.runtime.tree.ParseTree} representation of an expression while {@link org.antlr.v4.runtime.tree.ParseTreeWalker}
+ * traverses through the {@link org.antlr.v4.runtime.tree.ParseTree}.
+ *
+ * Use case:
+ * ParseTreeWalker walker = new ParseTreeWalker();
+ * ParseTreeEvaluatorListener listener = new ParseTreeEvaluatorListener(...);
+ * listener.initialize(...);
+ * walker.walk(listener, ...);
+ * final Object result = listener.getResult();
+ */
+@Named
+class ParseTreeEvaluatorListener implements DataPrepperExpressionListener {
+    private static final Set<Integer> CONDITIONAL_OPERATOR_TYPES = Set.of(AND, OR);
+    private static final Set<Integer> EQUALITY_OPERATOR_TYPES = Set.of(EQUAL, NOT_EQUAL);
+    private static final Set<Integer> RELATIONAL_OPERATOR_TYPES = Set.of(LT, LTE, GT, GTE);
+
+    private final Map<Integer, Operator<?>> strategy;
+    private final CoercionService coercionService;
+    private Stack<Integer> operatorSymbolStack = new Stack<>();
+    private Stack<Object> operandStack = new Stack<>();
+    private Event context;
+
+    @Inject
+    ParseTreeEvaluatorListener(final List<Operator<?>> operators, final CoercionService coercionService) {
+        this.coercionService = coercionService;
+        strategy = new HashMap<>(operators.size());
```

Comment:
The approach of taking a list of `Operator` from the application context and creating a map is a good approach. I do recommend that you abstract this functionality into its own class - `OperatorProvider`. It will be a small class, but very easy to test. And it can simplify the test for `ParseTreeEvaluatorListener`.

```
class OperatorProvider {

  private final Map<Integer, Operator<?>> codesToOperators;

  @Inject
  OperatorProvider(final List<Operator<?>> operators) {
    codesToOperators = new HashMap<>(operators.size());
    for (final Operator<?> operator : operators) {
        codesToOperators.put(operator.getSymbol(), operator);
    }
  }

  Operator<?> getOperator(Integer operatorCode) {
    Operator<?> operator = codesToOperators.get(operatorCode);
    if(operator == null) {
      throw ...
    }
  }
}
```